sealtk/core/test/KwiverTrackModel.cpp

```diff
@@ -0,0 +1,182 @@
+    auto const kbox = kv::bounding_box_d{{qbox.left(), qbox.top()},
+                                         qbox.width(), qbox.height()};
+
+    auto d = std::make_shared<kv::detected_object>(kbox);
+    auto s = std::make_shared<kv::object_track_state>(++frame, box.key(),
+                                                      std::move(d));
+
+    track->append(std::move(s));
+  }
+
+  return track;
+}
+
+// ----------------------------------------------------------------------------
+void testTrackData(QAbstractItemModel const& model, int row,
+                   kv::track_id_t id, TimeMap<QRectF> const& boxes)
+{
+  auto const& index = model.index(row, 0);
+
+  QVERIFY(model.data(index, LogicalIdentityRole).canConvert<kv::track_id_t>());
+  QCOMPARE(model.data(index, LogicalIdentityRole).value<kv::track_id_t>(), id);
+
+  QVERIFY(model.data(index, StartTimeRole).canConvert<time_us_t>());
+  QCOMPARE(model.data(model.index(row, 0), StartTimeRole).value<time_us_t>(),
+           boxes.firstKey());
+
+  QVERIFY(model.data(index, EndTimeRole).canConvert<time_us_t>());
+  QCOMPARE(model.data(model.index(row, 0), EndTimeRole).value<time_us_t>(),
+           boxes.lastKey());
+}
+
+} // namespace <anonymous>
+
+// ============================================================================
+class TestKwiverTrackModel : public QObject
+{
+  Q_OBJECT
+
+private slots:
+  void operations();
+  void data();
+};
+
+// ----------------------------------------------------------------------------
+void TestKwiverTrackModel::operations()
+{
+  KwiverTrackModel model;
+
+  QCOMPARE(model.rowCount(), 0);
+
+  auto const ts1 = std::make_shared<kv::object_track_set>(
+    std::vector<kv::track_sptr>{
+      createTrack(1, track1),
+    });
+
+  model.addTracks(ts1);
+  QCOMPARE(model.rowCount(), 1);
+
+  auto const ts2 = std::make_shared<kv::object_track_set>(
+    std::vector<kv::track_sptr>{
+      createTrack(2, track2),
+      createTrack(3, track3),
+    });
+
+  model.addTracks(ts2);
+  QCOMPARE(model.rowCount(), 3);
+
+  auto const ts3 = std::make_shared<kv::object_track_set>(
+    std::vector<kv::track_sptr>{
+      createTrack(4, track4),
+      createTrack(5, track5),
+    });
+
+  model.setTracks(ts3);
+  QCOMPARE(model.rowCount(), 2);
+
+  model.clear();
+  QCOMPARE(model.rowCount(), 0);
+}
+
+// ----------------------------------------------------------------------------
+void TestKwiverTrackModel::data()
+{
+  KwiverTrackModel model;
+  model.setTracks(
+    std::make_shared<kv::object_track_set>(
+      std::vector<kv::track_sptr>{
+        createTrack(1, track1),
+        createTrack(2, track2),
+        createTrack(3, track3),
+        createTrack(4, track4),
+        createTrack(5, track5),
+      }));
+  QCOMPARE(model.rowCount(), 5);
+
+  testTrackData(model, 0, 1, track1);
+  testTrackData(model, 1, 2, track2);
+  testTrackData(model, 2, 3, track3);
+  testTrackData(model, 3, 4, track4);
+  testTrackData(model, 4, 5, track5);
```

Comment:
Is there a reason this test can't be split into "data" and "test" parts the way other Qt tests are?

Reply:
What would be the effect of that, except possibly to move the definitions of `trackN` from an anonymous namespace to `operations_data()`? Is that worth the extra "frills"?

(Note: it is absolutely intended that the test tests all five tracks at once... the objective is to test a model with *multuple* tracks, not just to individually test different tracks.)

Personally I am inclined to only use the test/data split if I am going to run the same test code on more than one set of data. For that use case, it is definitely valuable. For a single data set, however, I'm not convinced it is worthwhile. (OTOH, if you have suggestions how this test would be improved by using multiple data sets...?)